### ISSUE TITLE
VXLAN/GRE/MPLS inner packet

### ISSUE THREAD

**Issue by RaidAndFade** [akvorado/akvorado#1445]
Hey, As title suggests, would be nice to be able to view inside of VXLAN/GRE/MPLS tunnels and see inner src/dests as well as possibly tunnel-specific facts such as mpls label / vxlan vni, which should all be visible in flows. 

Not sure how this would fit into current codebase... aside from obvious solution of adding secondary src/dest, naive suggestion is to provide a config option that reads inside of a tunnel and records only the inner data, and then some new field that records the tunnel type / flags

Thanks for your time!

**Comment by vincentbernat** [comment]
MPLS labels are already supported. In the case of MPLS, src/dst address are already the inner ones.

For the other types, it can be implemented as well for sFlow when you have a  sampled Ethernet or IP packet. Is that your case?

**Comment by RaidAndFade** [comment]
Yes, my case is arista VXLAN on Sflow, ideally with vni and any other data in the sample.

 Im not well versed on sflow format so unsure what the potential is but would be great to have the VNI and dest VTEP (outer ip) if possible. 

From what ive seen it might need some enrichment as ingress vxlan is sent as tunneled but egress seems like a regular packet sent to vxlan interface.

I could provide sample data if you provide steps

**Comment by vincentbernat** [comment]
Capture some of the sFlow packets you have for both directions (with tcpdump). With Wireshark, you can then export the ones that are interesting.

**Comment by RaidAndFade** [comment]
Sorry for the delay. Is it possible for you to provide me with a secure contact so I can send this data? Otherwise I would have to send obscured hexdumps.

I looked and did indeed find some interesting data in the samples, however it would be retrieved from the raw packet headers from what i can tell, there doesnt seem to be any sflow data to make it easier. But all of the data I want is there (The packet header is long enough to contain the VNI and includes the whole ethernet packet)

**Comment by RaidAndFade** [comment]
Looking deeper on both sides. Unfortunately it looks like arista only sends sflows for received vxlan packets indeed. 

Outbound vxlan's only indicator is that it has a different gateway next_hop than the router next_hop. No VNI information or any other indication of it being VXLAN.

**Comment by vincentbernat** [comment]
You can send a pcap to vincent@bernat.ch.

**Comment by RaidAndFade** [comment]
Sent, turns out i was wrong with the gateway being the destination vtep. it's actually the gateway *within* the vni, means no clear way to tie that to a vxlan. 

Let me know your thoughts and if we can do anything with tx, rx should be easy and should go a long way to clean my graphs since right now all vxlan incoming shows up as just vxlan...

**Comment by vincentbernat** [comment]
I suppose we could add an option to a listener to decap. You would send flows to decap to one listener, and flows to not decap to another. Would it work in your setup?

**Comment by vincentbernat** [comment]
Couldn't you configure a sampler for ingress (where decap happens) and one for egress? This will be far more generic than relying on some fields that may or may not be present.

**Comment by RaidAndFade** [comment]
ingress and egress both happen on every device, it would have to separate on a per device level and i dont know of any way to do that.

**Comment by RaidAndFade** [comment]
Additionally, knowing the VNI/VTEPs is critical to being able to use this data meaningfully. Without the VNI i may as well just strip the vxlan headers on the way in and rely on only the inner packets.

**Comment by vincentbernat** [comment]
I need a way that is generic enough to work on many users. The way I see it, you have interfaces where you expect VXLAN and interfaces where you don't expect VXLAN. On each interface, you can configure egress/ingress differently. That's why I was proposing you have two collectors. Specific fields can be collected nonetheless. But decapsulating if the field is present seems less generic (some implementations many never have the fields).

**Comment by RaidAndFade** [comment]
I understand, I wouldn't want this to be reimplemented due to specific platform limitations later on either. That's why I was thinking we could ignore any flow data and go strictly based on whether the internal raw packet header is a vxlan header or not, that way we would hopefully have a more platform agnostic method.

I don't believe I actually have the ability to send ingress/egress to different collectors on Arista EOS, it sends all configured interfaces to all configured collectors. Admittedly i'm only sampling on ingress right now on my sampled interfaces. Maybe some testing there on my end to see if setting "Egress" gives usable vxlan data, but i think thats just additional work not really getting to the core of the implementation.

I think it would be best implemented the way we have MACs now, where they are an optional configuration that is default disabled but can be enabled by the user. If your worry is bloat I believe that should solve the issue. 

Im open to any other ideas, however two collectors would probably not work in my instance, I still cant envision any situation where i would be able to separate ingress and egress on different collectors.

Note that I also don't really require the vxlan fields to ONLY be there if the agent is the dest vtep, i actually prefer to decap all vxlan packets regardless, as long as there is a queryable way to tell them apart (i.e. VxlanVni=0 in query)

**Comment by RaidAndFade** [comment]
I'm thinking out loud here but maybe down the line a way to specify which agents should decap ? a config file or some sort of flag we could include, that way the users that do not want to decap every vxlan but rather only some devices a way to whitelist? Or perhaps even as granular some "NoVXLAN" in the interface description to ignore ? I think we can do per-interface / per-device decap if we wanted to that way, though im not sure if that aligns with the project's goals. 

I feel once we get into this level of configurability it becomes quite hard to become generic for every use-case, a simple toggle on/off on decap seems ideal to me. No change for users who dont want it and users like me get to use it to its full extent.

**Comment by vincentbernat** [comment]
Decap any VXLAN does not seem robust as your own users could use VXLAN and we would decap them. Decap on the condition of some metadata could work. There is already the language used for classifiers, we could reuse it for that.
